src/dtx/dtx_common.c

```diff
@@ -0,0 +1,528 @@
+/**
+ * (C) Copyright 2019 Intel Corporation.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * GOVERNMENT LICENSE RIGHTS-OPEN SOURCE SOFTWARE
+ * The Government's rights to use, modify, reproduce, release, perform, display,
+ * or disclose this software are subject to the terms of the Apache License as
+ * provided in Contract No. B609815.
+ * Any reproduction of computer software, computer software documentation, or
+ * portions thereof marked with this legend must also reproduce the markings.
+ */
+/**
+ * dtx: DTX common logic
+ */
+#define D_LOGFAC	DD_FAC(dtx)
+
+#include <uuid/uuid.h>
+#include <abt.h>
+#include <daos_srv/pool.h>
+#include <daos_srv/container.h>
+#include <daos_srv/vos.h>
+#include <daos_srv/dtx_srv.h>
+#include <daos_srv/daos_server.h>
+#include "dtx_internal.h"
+
+struct dtx_batched_commit_args {
+	d_list_t		 dbca_link;
+	struct ds_cont_hdl	*dbca_cont_hdl;
+};
+
+void
+dtx_aggregate(void *arg)
+{
+	struct ds_cont		*cont = arg;
+
+	while (!cont->sc_closing) {
+		int	rc;
+
+		rc = vos_dtx_aggregate(cont->sc_hdl, DTX_AGG_YIELD_INTERVAL,
+				       DTX_AGG_THRESHOLD_AGE_LOWER);
+		if (rc != 0)
+			break;
+
+		ABT_thread_yield();
+	}
+
+	cont->sc_dtx_aggregating = 0;
+	ds_cont_put(cont);
+}
+
+static void
+dtx_flush_committable(struct ds_cont_hdl *cont_hdl)
+{
+	struct ds_cont		*cont = cont_hdl->sch_cont;
+	int			 rc;
+
+	do {
+		struct dtx_entry	*dte = NULL;
+
+		rc = vos_dtx_list_committable(cont->sc_hdl,
+					      DTX_THRESHOLD_COUNT, &dte);
+		if (rc <= 0)
+			break;
+
+		rc = dtx_commit(cont_hdl->sch_pool->spc_uuid, cont->sc_uuid,
+				dte, rc, cont_hdl->sch_pool->spc_map_version);
+		D_FREE(dte);
```

Comment:
I understand what you want, the symmetry ALLOC/FREE may makes the code more understandable, but it makes the code more redundant, because each user needs to call something like list_committable_count() firstly, then allocate the buffer, and then call list_committable() with such buffer, and then free the buffer after dtx_commit(). I prefer to make the logic to be shared as much as possible.

Reply:
I'm fine with the list_committable allocating the data...which is why I suggest something like 
get_committable instead of list_committable as you're not really creating a list anyway but a buffer of committable entries.

So really, I'm not sure I understand your comment.   

vos_dtx_get_committable(...) /* does what your list_committable does now */
vos_dtx_free_committable(...) /* just calls D_FREE unless at some later time we have some more complex allocation in which case it will do everything needed */

Reply:
OK, I will adjust the patch.